op-gateway/src/OPProofService.ts

```diff
@@ -15,58 +42,200 @@ const L2_OUTPUT_ORACLE_ABI = [
+  'function rootClaim() external view returns (bytes32 rootClaim_)',
+  // Status of the game challenging
+  'function status() external view returns (uint8)',
+];
+
 /**
  * The proofService class can be used to calculate proofs for a given target and slot on the Optimism Bedrock network.
  * It's also capable of proofing long types such as mappings or string by using all included slots in the proof.
  *
  */
 export class OPProofService implements IProofService<OPProvableBlock> {
-  private readonly l2OutputOracle: Contract;
+  private readonly l1Provider: JsonRpcProvider;
   private readonly l2Provider: JsonRpcProvider;
   private readonly helper: EVMProofHelper;
   private readonly delay: number;
 
+  private readonly optimismPortal: Contract;
+  private l2OutputOracle: Contract | undefined;
+  private disputeGameFactory: Contract | undefined;
+
   constructor(
     l1Provider: JsonRpcProvider,
     l2Provider: JsonRpcProvider,
-    l2OutputOracleAddress: string,
+    optimismPortalAddress: string,
     delay: number
   ) {
+    this.l1Provider = l1Provider;
     this.l2Provider = l2Provider;
     this.helper = new EVMProofHelper(l2Provider);
     this.delay = delay;
-    this.l2OutputOracle = new Contract(
-      l2OutputOracleAddress,
-      L2_OUTPUT_ORACLE_ABI,
+
+    this.optimismPortal = new Contract(
+      optimismPortalAddress,
+      OPTIMISM_PORTAL_ABI,
       l1Provider
     );
   }
 
+  async findDisputeGame(): Promise<OPProvableBlock | null> {
+    if (!this.disputeGameFactory) {
+      try {
+        const disputeGameFactoryAddress = await this.optimismPortal.disputeGameFactory();
+  
+        this.disputeGameFactory = new Contract(
+          disputeGameFactoryAddress,
+          DISPUTE_GAME_FACTORY_ABI,
+          this.l1Provider
+        );
+      } catch (err) {
+        console.error(err)
+        return null;
+      }
+    }
+
+    if (this.disputeGameFactory) {
+      const respectedGameType = await this.optimismPortal.respectedGameType();
+
+      /**
+       * Get the latest output from the L2Oracle. We're building the proof with this batch
+       * We go a few batches backwards to avoid errors like delays between nodes
+       *
+       */
+      const gameCount = Number(await this.disputeGameFactory.gameCount());
+
+      const games: FindLatestGamesResult[] =
+        await this.disputeGameFactory.findLatestGames(
+          respectedGameType,
+          gameCount - 1,
+          50
+        );
+
+      const timestampNow = (await this.l2Provider.getBlock('latest'))!
+        .timestamp;
+      let disputeGameIndex = -1;
+      let disputeGameLocalIndex = -1;
+
+      for (let i = 0; i < games.length; i++) {
+        if (timestampNow - Number(games[i].timestamp) >= this.delay) {
+          disputeGameIndex = Number(games[i].index);
+          disputeGameLocalIndex = i;
+          break;
+        }
+      }
+
+      // If game is not found then fallback to the L2OutputOracle
+      // for just merged case
```

Comment:
This would be easier to read if it returns null if `disputeGameIndex == -1`, making it possible to remove the if block below.